system/security/securesocket/securesocket.cpp

```diff
@@ -1177,105 +1284,123 @@ class CSecureSocketContext : implements ISecureSocketContext, public CInterface
-            throw MakeStringException(-1, "error loading certificate chain file %s - %s", certfile, errbuf);
+            throw makeStringExceptionV(-1, "error loading certificate chain file %s - %s", certFileorBuf, errbuf);
         }
+    }
+
+    void setPrivateKey(const char *privKeyFileorBuf)
+    {
+        if (isEmptyString(privKeyFileorBuf))
+            return;
 
-        if(SSL_CTX_use_PrivateKey_file(m_ctx, privkeyfile, SSL_FILETYPE_PEM) <= 0)
+        if (isPEMBuffer(privKeyFileorBuf))
+        {
+            // single RSA key in buffer
+            if (!usePrivateKeyPEMBuffer(m_ctx, privKeyFileorBuf))
+            {
+                char errbuf[512];
+                ERR_error_string_n(ERR_get_error(), errbuf, 512);
+                throw makeStringExceptionV(-1, "error loading private key %s", errbuf);
+            }
+        }
+        else if (SSL_CTX_use_PrivateKey_file(m_ctx, privKeyFileorBuf, SSL_FILETYPE_PEM) <= 0)
         {
             char errbuf[512];
             ERR_error_string_n(ERR_get_error(), errbuf, 512);
-            throw MakeStringException(-1, "error loading private key file %s - %s", privkeyfile, errbuf);
+            throw makeStringExceptionV(-1, "error loading private key file %s - %s", privKeyFileorBuf, errbuf);
         }
 
-        if(!SSL_CTX_check_private_key(m_ctx))
+        if (!SSL_CTX_check_private_key(m_ctx))
+            throw makeStringExceptionV(-1, "Private key does not match the certificate public key");
+    }
+
+    void setVerifyLocations(const char *caCertsPathorBuf)
+    {
+        if (isEmptyString(caCertsPathorBuf))
+            return;
+
+        if (isPEMBuffer(caCertsPathorBuf))
         {
-            throw MakeStringException(-1, "Private key does not match the certificate public key");
+            // can have multiple certs in buffer
+            if (!loadVerifyLocationsPEMBuffer(m_ctx, caCertsPathorBuf))
+                throw makeStringExceptionV(-1, "Error loading CA certificates");
         }
-        
+        else if (SSL_CTX_load_verify_locations(m_ctx, caCertsPathorBuf, NULL) != 1)
+            throw makeStringExceptionV(-1, "Error loading CA certificates from %s", caCertsPathorBuf);
+    }
+
+public:
+    CSecureSocketContext(SecureSocketType sockettype)
+    {
+        initContext(sockettype);
+
         SSL_CTX_set_mode(m_ctx, SSL_CTX_get_mode(m_ctx) | SSL_MODE_AUTO_RETRY);
     }
 
-    CSecureSocketContext(const IPropertyTree* config, SecureSocketType sockettype)
+    CSecureSocketContext(const char* certFileorBuf, const char* privKeyFileorBuf, const char* passphrase, SecureSocketType sockettype)
     {
-        assertex(config);
-        m_verify = false;
-        m_address_match = false;
+        initContext(sockettype);
 
-        if(sockettype == ClientSocket)
-            m_meth = SSLv23_client_method();
-        else
-            m_meth = SSLv23_server_method();
+        // MCK TODO: should we set a default cipherList, as is done in other ctor (below) ?
```

Comment:
Should we an explicit cipherList as is done in other ctor ?

Reply:
looks like we should provide a way for an opional cipherList to be provided via this ctor route too,
and default like the other ctor if not.
and/or could it a method of ISecureSocketContext and (optionally) called to set after construction?